apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/chat/ChatView.kt

```diff
@@ -961,21 +989,104 @@ fun BoxScope.ChatItemsList(
+
+  LaunchedEffect(Unit) {
+    launch {
+      snapshotFlow { chatModel.chatId.value }
+        .distinctUntilChanged()
+        .collect {
+          if (revealedItems.value.isNotEmpty()) {
+            revealedItems.value = setOf()
+          }
+          preloadItemsEnabled.value = true
+          val firstUnreadItem = reversedChatItems.value.findLast { it.isRcvNew }
+          if (firstUnreadItem != null) {
+            val firstUnreadItemIndexIdx = sections.chatItemPosition(firstUnreadItem.id)
+            if (firstUnreadItemIndexIdx != null) {
+              listState.scrollToItem(scrollPosition(firstUnreadItemIndexIdx), -maxHeight.value)
+            }
+
+            if (chatModel.chatItemsSectionArea[firstUnreadItem.id] != ChatSectionArea.Bottom) {
+              withBGApi {
+                scrollAdjustmentEnabled.value = false
+                try {
+                  apiLoadBottomSection(chatInfo, chatModel, remoteHostId)
+                } finally {
+                  delay(600)
+                  scrollAdjustmentEnabled.value = true
+                }
+              }
+            }
+          }
         }
+    }
+  }
+
+  val scrollToItem: State<(Long) -> Unit> = remember { mutableStateOf({ itemId: Long ->
+    val index = sections.chatItemPosition(itemId)
+    preloadItemsEnabled.value = false
+
+    if (index != null) {
+      scope.launch {
+        listState.animateScrollToItem(scrollPosition(index), -maxHeight.value)
+        preloadItemsEnabled.value = true
       }
-    )
+    } else {
+      withBGApi {
+        try {
+          val destinationSection = sections.find { it.boundary.area == ChatSectionArea.Destination }
+          val itemsToDrop = destinationSection?.items?.flatMap { it.items }?.toList()
+          withContext(Dispatchers.Main) {
+            itemsToDrop?.forEach {
+              chatModel.chatItemsSectionArea[it.id] = ChatSectionArea.Current
+            }
+          }
+          val chatSectionLoader = ChatSectionLoader(0, ChatSectionArea.Destination)
+          apiLoadMessages(
+            rhId = remoteHostId,
+            chatInfo = chatInfo,
+            chatModel = chatModel,
+            itemId = itemId,
+            search = "",
+            chatSectionLoader = chatSectionLoader,
+            pagination = ChatPagination.Around(itemId, ChatPagination.PRELOAD_COUNT * 2)
+          )
+
+          val idx = sections.chatItemPosition(itemId)
+          scope.launch {
+            if (idx != null) {
+              listState.animateScrollToItem(scrollPosition(idx), -maxHeight.value)
+              withContext(Dispatchers.Main) {
+                if (!itemsToDrop.isNullOrEmpty()) {
+                  itemsToDrop.forEach {
+                    chatModel.chatItemsSectionArea.remove(it.id)
+                  }
+                  chatModel.chatItems.value.removeIf { chatModel.chatItemsSectionArea[it.id] == null }
```

Comment:
`removeif` is Java function. Kotlin equivalent is `.removeAll`.  In my practice I remember that when I used it on some old Android version i got crash. Maybe it's lower API when we support (like 21, for example), but anyway, better to use Kotlin functions that do the same as Java functions to be on a safe side

Reply:
I also noticed that I used it in different place, nice. So it will not make a crash on our min sdk level, but anyway

Reply:
Also, never modify chatModel.chats.value directly like you do with .removeIf. It should always be non-modifyable list, only items at index can change, not the elements order/size. Apply your changes to a new list and set this list to chatItems using the following block:

```kotlin
withChats {
    chats.removeAll {  }
}
```

Here `withChats` prevents concurrency and gives you chats instance you can use for modification using implemented helpers that will NOT modify the list but create a new one and re-set it to chatModel.chatItems.

So .removeAll here is a helper written in ChatModel.kt:2300, take a look.

The idea of why we do this strange thing is to prevent issues with concurrent modification of data that gave us a lot of random problems with LazyColumn in the past.